cdap-app-templates/cdap-program-report/src/main/scala/co/cask/cdap/report/ReportGenerationHelper.scala

```diff
@@ -0,0 +1,185 @@
+    * |  required column 1  | required column 2  | required columns ...
+    * +---------------------+--------------------+----------------------
+    * For instance, if the required columns only contain three columns "namespace", "program", and "run",
+    * the final [[org.apache.spark.sql.DataFrame]] will contain columns as shown below:
+    * +---------+---------------+-----------+
+    * |   run   |   namespace   |  program  |
+    * +---------+---------------+-----------+
+    * The final [[org.apache.spark.sql.DataFrame]] will be written to a JSON file at the given output location,
+    * accompanied by an empty _SUCCESS file indicating success.
+    *
+    * @param spark the spark session to run report generation with
+    * @param request the request containing the requirement for the report generation
+    * @param inputURIs URIs of the avro files containing program run meta records
+    * @param outputLocation location of the output directory where the report file and _SUCCESS file will be written
+    * @throws java.io.IOException when fails to write to the _SUCCESS file
+    */
+  @throws(classOf[IOException])
+  def generateReport(spark: SparkSession, request: ReportGenerationRequest, inputURIs: java.util.List[String],
+                     outputLocation: Location): Unit = {
+    import spark.implicits._
+    val df = spark.read.format("com.databricks.spark.avro").load(inputURIs: _*)
+    // Group the program run meta records by program run Id's and aggregate grouped records into a column
+    // with data type Record. The aggregated DataFrame aggDf will have two columns: "run" and "record"
+    val aggCol = new RecordAggregator().toColumn.alias(RECORD_COL).as[Record]
+    // TODO: configure partitions. The default number of partitions is 200
+    var aggDf = df.groupBy(Constants.RUN).agg(aggCol)
+    // Construct a set of fields to be included in the final report with required fields and fields from the request
+    val reportFields: collection.mutable.LinkedHashSet[String] =
+    collection.mutable.LinkedHashSet(REQUIRED_FIELDS: _*) ++ Option (request.getFields)
+      .getOrElse[java.util.List[String]](Collections.emptyList[String]())
+    LOG.debug("Fields to be included in the report: {}", reportFields)
+    // Create a set of additional fields to be included as columns in the aggregated DataFrame
+    // Initialize the set with "start" and "end" for filtering records according to the time range [start, end)
+    // specified in the request
+    var additionalFields = collection.mutable.LinkedHashSet(Constants.START, Constants.END)
+    // Add field names for filtering
+    additionalFields ++= Option(request.getFilters)
+      .getOrElse[java.util.List[Filter[_]]](Collections.emptyList[Filter[_]]()).map(f => f.getFieldName)
+    // Add field names for sorting
+    additionalFields ++= Option(request.getSort)
+      .getOrElse[java.util.List[Sort]](Collections.emptyList[Sort]()).map(s => s.getFieldName)
```

Comment:
Use `import` instead of always using full class name. You can also do `import java.util.{List => JList}` if there is name clash between Java collection and scala one.